src/xrpld/app/paths/detail/MPTEndpointStep.cpp

```diff
@@ -0,0 +1,1056 @@
+    {
+        // Between holders
+        if (isDirectBetweenHolders_)
+        {
+            auto const& holder = ctx.isFirst ? src_ : dst_;
+            // Payment between the holders
+            if (isFrozen(ctx.view, holder, mptIssue_))
+                return tecLOCKED;
+
+            if (auto const ter =
+                    canTransfer(ctx.view, mptIssue_, holder, ctx.strandDst);
+                ter != tesSUCCESS)
+                return ter;
+        }
+        // Don't need to check if a payment is between issuer and holder
+        // in either direction
+    }
+    // Cross-token MPT payment via DEX
+    else
+    {
+        auto const account = ctx.isFirst ? src_ : dst_;
+        if (auto const ter = isMPTDEXAllowed(
+                ctx.view, mptIssue_, account, mptIssue_.getIssuer());
+            ter != tesSUCCESS)
+            return ter;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+MPTEndpointOfferCrossingStep::check(
+    StrandContext const& ctx,
+    std::shared_ptr<const SLE> const&) const
+{
+    auto const& holder = ctx.isFirst ? src_ : dst_;
+    auto const& issuer = mptIssue_.getIssuer();
+    if (holder != issuer)
+    {
+        if (auto const ter =
+                isMPTDEXAllowed(ctx.view, mptIssue_, holder, issuer);
+            ter != tesSUCCESS)
+            return ter;
+    }
+    return tesSUCCESS;
+}
+
+TER
+MPTEndpointOfferCrossingStep::checkCreateMPT(
+    ApplyView& view,
+    ripple::DebtDirection srcDebtDir)
+{
+    // TakerPays is the last step if offer crossing
+    if (isLast_)
+    {
+        // Create MPToken for the offer's owner. No need to check
+        // for the reserve since the offer doesn't go on the books
+        // if crossed. Insufficient reserve is allowed if the offer
+        // crossed. See CreateOffer::applyGuts() for reserve check.
+        if (auto const err =
+                MPTokenAuthorize::checkCreateMPT(view, mptIssue_, dst_, j_);
+            err != tesSUCCESS)
+        {
+            JLOG(j_.trace())
+                << "MPTEndpointStep::checkCreateMPT: failed create MPT";
+            resetCache(srcDebtDir);
+            return err;
+        }
+    }
+    return tesSUCCESS;
+}
+
+//------------------------------------------------------------------------------
+
+template <class TDerived>
+std::pair<MPTAmount, DebtDirection>
+MPTEndpointStep<TDerived>::maxPaymentFlow(ReadView const& sb) const
+{
+    // From a holder to an issuer
+    if (src_ != mptIssue_.getIssuer())
+        return {
+            toAmount<MPTAmount>(accountHolds(
+                sb, src_, mptIssue_, fhIGNORE_FREEZE, ahIGNORE_AUTH, j_)),
+            DebtDirection::redeems};
+
+    // From an issuer to a holder
+    if (auto const sle = sb.read(keylet::mptIssuance(mptIssue_)))
+    {
+        std::uint64_t const maximumAmount = [&] {
+            auto const max = sle->getFieldU64(sfMaximumAmount);
+            return max > 0 ? max : maxMPTokenAmount;
```

Comment:
`sfMaximumAmount` is defined as `soeOPTIONAL` (not `soeDEFAULT` ), so I'm not sure if the default will be zero. But for best practice I would propose return `(*sle)[~sfMaximumAmount].value_or(maxMPTokenAmount)`